src/QirRuntime/lib/Simulators/ToffoliSimulator.cpp

```diff
@@ -67,21 +65,23 @@ namespace Quantum
         Qubit AllocateQubit() override
         {
             this->lastUsedId++;
-            this->states.ExtendToInclude(this->lastUsedId);
+            this->states.emplace_back();
             return reinterpret_cast<Qubit>(this->lastUsedId);
         }
 
         void ReleaseQubit(Qubit qubit) override
         {
             const long id = GetQubitId(qubit);
             assert(id <= this->lastUsedId);
-            assert(!this->states.IsBitSetAt(id));
+            assert(!this->states.at(id));
+            this->lastUsedId--;
+            this->states.pop_back();
         }

```

Comment:
I have an impression that this function works correctly for the _last-most qubit id only_. But if we release a qubit "in the middle" then still the last-most qubit will be released. E.g. we call `AllocateQubit()` 3 times. It returns 0, 1, 2. Then we call `ReleaseQubit(0)`, the fragment `this->states.pop_back()` removes the qubit with id 2 (rather than with id 0).

Reply:
Yes, that's correct, but in this case I'm relying on Q#'s strict qubit scoping for a shortcut. In Q#, every qubit has a scope, either implied or explicit, such that any qubit allocations are always released in the reverse order they were allocated. This does mean that the ToffoliSimulator could exhibit the bug you describe if it is used for QIR that was generated from something other than Q# (or written manually). However, because the simulators are packages as part of `Microsoft.Quantum.Qir.QSharp.Core.dll` it seems reasonable to restrict their behavior to Q# generated QIR. If someone wanted to use these simulators for another language, they'd have to conform to Q# gateset AND Q#'s lifetime semantics, which seems reasonable to me.

That said, this is a very subtle requirement that might not be obvious, and I'm open to further discussion. @bettinaheim, any concerns or suggestions about how to handle this?

Reply:
If the qubits are always deallocated in the reverse order of allocation then we are fine here. Can mark this conversation as resolved.

Reply:
In that case the parameter (`Qubit qubit`) becomes redundant and misleads:
```c++
void ReleaseQubit(Qubit qubit) override
```